gammapy/irf/core.py

```diff
@@ -45,6 +45,9 @@ def __init__(self, axes, data=0, unit="", meta=None, interp_kwargs=None):
         if isinstance(data, u.Quantity):
             self.data = data.value
             self.unit = data.unit
+            if self.axes.names == ["energy_true", "migra", "offset"] :
+                if self.unit != u.dimensionless_unscaled:
+                    raise ValueError(f"Error: {self.unit} is not an allowed unit. EDisp requires dimensionless data quantities!")
```

Comment:
So I guess None of the IRFs specified units?

Then we should just add a `unit` attribute to all of the classes and use that here to check using:
```
if not self.unit.is_equivalent(data.unit):
    raise ValueError(...)
```

Reply:
Yes, this should be done in a general way. There is already a `IRF.unit` attribute, which specifies the currently used unit, bot not the expected one. I think the most elegant solution would be introduce a class level attribute `.default_unit` or similar, which specifies the expected unit equivalency for the data.

Reply:
Thanks @maxnoe @adonath , I added the default units for all the IRF extensions (Edisp, Aeff, BKG and PSF) and I also modified the tests accordingly.
Unfortunately, I've read the last message only after pushing so I called the default units simply as `unit`...

Reply:
Yes, please adapt the name to `.default_unit` and add the check like `u.Unit(val).is_equivalent(self.default_unit)` here: https://github.com/gammapy/gammapy/blob/master/gammapy/irf/core.py#L145

Reply:
Shouldn't that setter transform any existing data?

At the moment (as is often the case when having this getter/setter setups) the user can easily create invalid data like this:

```
aeff = EffectiveArea2D(data=[[1, 2, 3, 4]] * u.m**2)
aeff.unit = u.cm**2  # I would expect that to *convert*, not *overwrite*
```

Reply:
Good point. Thinking about it again we should probably remove the setter completely? This is the behavior for `Quantity`:
```
from astropy import units as u
q = 5 * u.TeV
q.unit = "GeV"
```
Raises `AttributeError: can't set attribute`, which makes sense to me...

Reply:
Instead for maps we have: https://github.com/gammapy/gammapy/blob/master/gammapy/maps/core.py#L1506 Which mirrors a bit the `Quantity.to()` API.

Reply:
Sorry,  it's not completely clear to me what's the final suggestion here, i.e. if I have to modify the `unit(self, val)` introducing some filters or not 😅.
For now, I've just implemented everywhere the simplest change (i.e. `default_unit`).

Reply:
The use cases to actually change the unit are more than rare, but it might still be reasonable to support it...

Reply:
@fabiopintore I think for this PR we probably keep as is, but we might need a follow up discussion to solve the question whether units should be settable at all (for `IRF` as well as `Map`). Right now they are which is fine, as it still a well define behavior. But I agree with @maxnoe that it is maybe unexpected and could cause more harm than benefit (especially because it is different behavior than for `Quantity`)

Reply:
If you *really* want to do it, you could do `map.quantity = u.Quantity(map.data, u.unit)` but then it's also clear that this is kind of dangerous.

So +1 for removing the unit setter (and probably even the data setter).